File: 27-03-2024/Find shortest safe route in a matrix.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution
{
public:
    int findShortestPath(vector<vector<int>> &mat)
    {
      //Your Code Start
         int n = mat.size();
       int m = mat[0].size();
       vector<vector<int>> vis(n , vector<int> (m , 0));
       vector<vector<int>> dis(n , vector<int> (m , 1e9));
       int minDis = 1e9;
       priority_queue <pair<int , pair<int , int>>, vector<pair<int , pair<int , int>>>, greater<pair<int , pair<int , int>>> > pq; 
       
       for(int i=0 ; i<n ; i++){
           for(int j=0 ; j<m ; j++){
               
               if(mat[i][j] == 0){
                   
                   vis[i][j] = 1;
                   
                   if(i-1 >= 0) vis[i-1][j] = 1;
                   if(j-1 >= 0) vis[i][j-1] = 1;
                   if(i+1 < n) vis[i+1][j] = 1;
                   if(j+1 < m) vis[i][j+1] = 1;
               }
           }
       }
       
       for(int i=0 ; i<n ; i++){
           
           if(!vis[i][0]){
              pq.push({1 , {i , 0}});
              dis[i][0] = 0;
           }
       }
       
       int dr[] = {-1 , 0 , 1 , 0};
       int dc[] = {0 , 1 , 0 , -1};
       
       while(!pq.empty()){
           
           int r = pq.top().second.first;
           int c = pq.top().second.second;
           int steps = pq.top().first;
           pq.pop();
           
           for(int i=0 ; i<4 ; i++){
               
               int nr = r + dr[i];
               int nc = c + dc[i];
               
               if(nr>=0 && nr<n && nc>=0 && nc<m && !vis[nr][nc]){
                   
                   if(dis[nr][nc] > 1 + steps){
                      dis[nr][nc] = 1 + steps; 
                      pq.push({steps+1 , {nr , nc}});
                   }
               }
           }
           
       }
       
       for(int i=0 ; i<n ; i++){
           
           minDis = min(dis[i][m-1] , minDis);
       }
       
       if(minDis != 1e9) return minDis;
       
       return -1;
       // Your Code End
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int R, C;
        cin >> R >> C;
        vector<vector<int>> mat(R, vector<int>(C));
        for (int i = 0; i < R; i++)
        {
            for (int j = 0; j < C; j++)
            {
                cin >> mat[i][j];
            }
        }

        Solution ob;
        int ans = ob.findShortestPath(mat);
        cout << ans << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: 27-03-2024/Find shortest safe route in a matrix.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split("\n").map(string => {
        return string.trim();
    });

    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++){
        const [r, c] = readLine().split(' ').map(x => parseInt(x));

        let mat = [];
        for(let i=0;i<r;i++)
        {
            let a = readLine().trim().split(' ').map((x) => parseFloat(x));
            mat.push(a);
        }
        
        let obj = new Solution();
        let res = obj.findShortestPath(r, c, mat);
        console.log(res);
    }
}

// } Driver Code Ends

class Solution {
    /**
     * @param {number} r
     * @param {number} c
     * @param {number[][]} mat
     * @returns {number}
     */
    findShortestPath(r, c, mat) {
      //Your Code Start
        const delrow = [-1, 0, 1, 0];
        const delcol = [0, 1, 0, -1];
        
        // Create a matrix to mark visited cells and obstacles
        const ans = [];
        for (let i = 0; i < r; i++) {
            ans.push(new Array(c).fill(1));
        }
        
        // Mark obstacles and adjacent cells as unreachable
        for (let i = 0; i < r; i++) {
            for (let j = 0; j < c; j++) {
                if (mat[i][j] === 0) {
                    ans[i][j] = 0;
                    for (let k = 0; k < 4; k++) {
                        const nr = delrow[k] + i;
                        const nc = delcol[k] + j;
                        if (nr >= 0 && nr < r && nc >= 0 && nc < c) {
                            ans[nr][nc] = 0;
                        }
                    }
                }
            }
        }
        
        // BFS traversal
        const queue = [];
        for (let i = 0; i < r; i++) {
            if (ans[i][0] === 1) {
                queue.push([i, 0, 1]);
            }
        }
        
        while (queue.length > 0) {
            const [row, col, count] = queue.shift();
            if (col === c - 1) {
                return count;
            }
            for (let i = 0; i < 4; i++) {
                const nr = delrow[i] + row;
                const nc = delcol[i] + col;
                if (nr >= 0 && nr < r && nc >= 0 && nc < c && ans[nr][nc] === 1) {
                    ans[nr][nc] = 0; // Mark as visited
                    queue.push([nr, nc, count + 1]);
                }
            }
        }
        
        return -1; // No path found
 //Your Code End
    }
}
